keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,948 @@
+start logging and generating events about the configuration mismatch.
+
+#### IPAddress
+
+
+The ServiceIPRange will act as a generator of IPAddresses, that will be defined as:
+
+```go
+// IPAddress represents an IP used by Kubernetes associated to an ServiceIPRange.
+// The name of the object is the IP address in canonical format.
+type IPAddress struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   IPAddressSpec   `json:"spec,omitempty"`
+	Status IPAddressStatus `json:"status,omitempty"`
+}
+
+// IPAddressSpec describe the attributes in an IP Address,
+type IPAddressSpec struct {
+	// ServiceIPRangeRef is a reference to the ServiceIPRange referenced
+  // +required
+	ServiceIPRangeRef *ServiceIPRangeReference
+}
+
+// ServiceIPRangeReference represents an ServiceIPRange reference
+type ServiceIPRangeReference struct {
+	// Name is unique to reference an ServiceIPRange resource.
+	Name string
+}
+```
+
+The IPAddress will have the following operations:
+
+- Create: 
+   - validate the name is an IP address in its canonical format
+   - add an annotation with the timestamp it was created, this is required because the IPAddress is
+   allocated before the Service is created and, if the apiserver crash during the operation, the
+   IPaddress will be orphan, but the controller doesn't have a way to know if it is really orphan or
+   just being part of a new Service created.
+   - add a finalizer
+
+- Update: immutable
+
+- Delete: an IP address controller will participate on the deletion removing the finalizer to avoid
+  that there are Services without an IP address object associated.
+
+
+The IPAddress object will be reconciled against the ServiceIPRanges and Services to guarantee that
+all IPAddresses created belong to an existing ServiceIPRange and that they are owned by an existing Service.
+
+The controller will avoid leaks cleaning all IPAddress without a Service owner reference or, if it doesn't has
+an owner reference, the time since it was created (it has annotation with the timestamp creation) is greater
+than 15 seconds.
+
+
+### Allocator
+
+A new allocator will be implemented that implements the current Allocator interface in the apiserver.
+
+```go
+// Interface manages the allocation of IP addresses out of a range. Interface
+// should be threadsafe.
+type Interface interface {
+	Allocate(net.IP) error
+	AllocateNext() (net.IP, error)
+	Release(net.IP) error
+	ForEach(func(net.IP))
+	CIDR() net.IPNet
+	IPFamily() api.IPFamily
+	Has(ip net.IP) bool
+	Destroy()
+	EnableMetrics()
+
+	// DryRun offers a way to try operations without persisting them.
+	DryRun() Interface
+}
+```
+
+This allocator will have an informer watching ServiceIPRanges and IPAddresses, so it can have locally the
+information needed to assign new IP addresses to Services.
+
+The uniqueness of an IPAddress is guaranteed by the IPAddress object, since trying two allocate a new IP
+address already exist will fail.
```

Comment:
```suggestion
The uniqueness of an IPAddress is guaranteed by the apiserver, since trying to allocate an IP
address that already exist will fail.
```